controller/app.go

```diff
@@ -191,69 +209,39 @@ func (r *AppRepo) List() (interface{}, error) {
-	query := "SELECT event_id, app_id, object_id, object_type, data, created_at FROM app_events WHERE app_id = $1 AND event_id > $2"
-	args := []interface{}{appID, sinceID}
-	n := 3
-	if typ != "" {
-		query += fmt.Sprintf(" AND object_type = $%d", n)
-		n++
-		args = append(args, typ)
-	}
-	if objectID != "" {
-		query += fmt.Sprintf(" AND object_id = $%d", n)
-		args = append(args, objectID)
-	}
-	query += " ORDER BY event_id DESC"
-	if count > 0 {
-		query += fmt.Sprintf(" LIMIT $%d", n)
-		args = append(args, count)
-	}
-	rows, err := r.db.Query(query, args...)
-	if err != nil {
-		return nil, err
-	}
-	var events []*ct.AppEvent
-	for rows.Next() {
-		event, err := scanAppEvent(rows)
-		if err != nil {
-			rows.Close()
-			return nil, err
-		}
-		events = append(events, event)
+func (r *AppRepo) SetRelease(appID string, releaseID string) error {
+	if err := r.db.Exec("UPDATE apps SET release_id = $2, updated_at = now() WHERE app_id = $1", appID, releaseID); err != nil {
+		return err
 	}
-	return events, nil
+	defer r.createEvent(appID, ct.AppEvent{
+		Type: "update",
+	})
+	return nil
 }
 
-func (r *AppRepo) GetEvent(id int64) (*ct.AppEvent, error) {
-	row := r.db.QueryRow("SELECT event_id, app_id, object_id, object_type, data, created_at FROM app_events WHERE event_id = $1", id)
-	return scanAppEvent(row)
+func (r *AppRepo) GetRelease(id string) (*ct.Release, error) {
+	row := r.db.QueryRow("SELECT r.release_id, r.artifact_id, r.data, r.created_at FROM apps a JOIN releases r USING (release_id) WHERE a.app_id = $1", id)
+	return scanRelease(row)
 }
 
-func scanAppEvent(s postgres.Scanner) (*ct.AppEvent, error) {
-	var event ct.AppEvent
-	var typ string
-	var data []byte
-	err := s.Scan(&event.ID, &event.AppID, &event.ObjectID, &typ, &data, &event.CreatedAt)
+func (r *AppRepo) createEvent(appID string, e ct.AppEvent) error {
+	app, err := selectApp(r.db, appID, false)
 	if err != nil {
-		if err == sql.ErrNoRows {
-			err = ErrNotFound
-		}
-		return nil, err
+		fmt.Printf("AppRepo.createEvent error (selectApp): %#v\n", err)
+		return err
 	}
-	event.AppID = postgres.CleanUUID(event.AppID)
-	event.ObjectType = ct.EventType(typ)
-	event.Data = json.RawMessage(data)
-	return &event, nil
-}
-
-func (r *AppRepo) SetRelease(appID string, releaseID string) error {
-	return r.db.Exec("UPDATE apps SET release_id = $2, updated_at = now() WHERE app_id = $1", appID, releaseID)
-}
-
-func (r *AppRepo) GetRelease(id string) (*ct.Release, error) {
-	row := r.db.QueryRow("SELECT r.release_id, r.artifact_id, r.data, r.created_at FROM apps a JOIN releases r USING (release_id) WHERE a.app_id = $1", id)
-	return scanRelease(row)
+	e.App = app
+	data, err := json.Marshal(e)
+	if err != nil {
+		fmt.Printf("AppRepo.createEvent error (json.Marshal): %#v\n", err)
+		return err
+	}
+	query := "INSERT INTO events (app_id, object_id, object_type, data) VALUES ($1, $2, $3, $4)"
+	if err := r.db.Exec(query, appID, appID, string(ct.EventTypeApp), data); err != nil {
```

Comment:
Given these events all follow DB actions, I wonder whether we should be creating the events using DB triggers (e.g. [like this](https://github.com/flynn/flynn/blob/master/controller/schema.go#L80))?